pkg/controller/nodeipam/ipam/clustercidrconfig_controller.go

```diff
@@ -0,0 +1,484 @@
+	//
+	// 5ms, 10ms, 20ms, 40ms, 80ms, 160ms, 320ms, 640ms, 1.3s, 2.6s, 5.1s, 10.2s, 20.4s, 41s, 82s
+	maxRetries                   = 15
+	defaultClusterCIDRConfigName = "default-cluster-cidr-config"
+)
+
+// ClusterCIDRConfigController reconciles ClusterCIDRConfig objects.
+type ClusterCIDRConfigController struct {
+	kubeClient clientset.Interface
+	// Recorder to record K8S events created by the controller.
+	recorder record.EventRecorder
+	lock     *sync.Mutex
+
+	clusterCIDRConfigLister networkinglisters.ClusterCIDRConfigLister
+	clusterCIDRConfigSynced cache.InformerSynced
+
+	queue workqueue.RateLimitingInterface
+
+	// CIDRMap keeps a mapping between the label and the associated cluster CIDRs.
+	CIDRMap map[string][]*cidrset.ClusterCIDR
+}
+
+func NewClusterCIDRConfigController(
+	clusterCIDRConfigInformer networkinginformers.ClusterCIDRConfigInformer,
+	kubeClient clientset.Interface,
+	allocatorParams CIDRAllocatorParams) (*ClusterCIDRConfigController, error) {
+	if kubeClient == nil {
+		klog.Fatalf("kubeClient is nil when starting Controller")
+	}
+
+	eventBroadcaster := record.NewBroadcaster()
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, v1.EventSource{
+		Component: "clusterCIDRConfigController",
+	})
+	eventBroadcaster.StartStructuredLogging(0)
+
+	klog.Infof("Sending events to api server.")
+	eventBroadcaster.StartRecordingToSink(
+		&v1core.EventSinkImpl{
+			Interface: kubeClient.CoreV1().Events(""),
+		})
+
+	if kubeClient.CoreV1().RESTClient().GetRateLimiter() != nil {
+		ratelimiter.RegisterMetricAndTrackRateLimiterUsage("cluster_cidr_config_controller", kubeClient.CoreV1().RESTClient().GetRateLimiter())
+	}
+
+	cccController := &ClusterCIDRConfigController{
+		kubeClient: kubeClient,
+		CIDRMap:    allocatorParams.CIDRMap,
+		recorder:   recorder,
+		lock:       allocatorParams.CIDRLock,
+		queue:      workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "cluster_cidr_config_controller"),
+	}
+
+	existingConfigList, err := listClusterCIDRConfigs(kubeClient)
+	if err != nil {
+		return nil, err
+	}
+
+	createDefaultClusterCIDRConfig(existingConfigList, allocatorParams)
+
+	// Regenerate the CIDRMaps from the existing ClusterCIDRConfigs.
+	for _, clusterCIDRConfig := range existingConfigList.Items {
+		klog.Infof("Creating ClusterCIDRConfig: %s", clusterCIDRConfig.Name)
+		// Create an event for invalid ClusterCIDRConfigs, do not crash on failures.
+		if err := cccController.reconcileBootstrap(&clusterCIDRConfig); err != nil {
+			klog.Errorf("Error while creating ClusterCIDRConfig: %v", err)
+			cccController.recorder.Event(&clusterCIDRConfig, "Warning", "InvalidClusterCIDRConfig", err.Error())
+		}
+	}
+
+	clusterCIDRConfigInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    cccutil.CreateClusterCIDRConfigHandler(cccController.reconcileCreate),
+		DeleteFunc: cccutil.CreateClusterCIDRConfigHandler(cccController.reconcileDelete),
+	})
```

Comment:
it seems you've created an event triggered controller, so you process the event directly in the handlers.
We use to implement level based controllers, so when the event is received, you enqueue it and then in the `syncClusterCIDRConfig` you have to get current status and "move" the cluster to the desired status, in this case that means to get all the nodes too ... we need to watch node events too, right?

Reply:
duh, there are 2 controllers, I've found the other now ... still my comment about event triggered stands, IIUIC this process the add and delete events in the handlers, objects are added to the queue only in `handleErr()`